sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -534,6 +535,18 @@ class DataSourceV2Strategy(session: SparkSession) extends Strategy with Predicat
       }
       UncacheTableExec(r.table, cascade = !isTempView(r.table)) :: Nil
 
+    case a @ AddCheckConstraint(PhysicalOperation(_, _, d: DataSourceV2ScanRelation), check) =>
```

Comment:
Can the optimizer replace `DataSourceV2ScanRelation` with `LocalRelation` and we won't have a match here?

Reply:
Good point. Created https://issues.apache.org/jira/browse/SPARK-52140 for this